tessera-core/src/main/java/com/quorum/tessera/transaction/ResendManagerImpl.java

```diff
@@ -0,0 +1,85 @@
+package com.quorum.tessera.transaction;
+
+import com.quorum.tessera.enclave.Enclave;
+import com.quorum.tessera.enclave.EncodedPayload;
+import com.quorum.tessera.enclave.PayloadEncoder;
+import com.quorum.tessera.enclave.model.MessageHash;
+import com.quorum.tessera.enclave.model.MessageHashFactory;
+import com.quorum.tessera.encryption.PublicKey;
+import com.quorum.tessera.transaction.model.EncryptedTransaction;
+
+import javax.transaction.Transactional;
+import java.util.Optional;
+
+public class ResendManagerImpl implements ResendManager {
+
+    private final EncryptedTransactionDAO encryptedTransactionDAO;
+
+    private final PayloadEncoder payloadEncoder;
+
+    private final Enclave enclave;
+
+    private final MessageHashFactory messageHashFactory = MessageHashFactory.create();
+
+    public ResendManagerImpl(final EncryptedTransactionDAO dao, final PayloadEncoder encoder, final Enclave enclave) {
+        this.encryptedTransactionDAO = dao;
+        this.payloadEncoder = encoder;
+        this.enclave = enclave;
+    }
+
+    //TODO: synchronize based on messagehash, so different message don't lock each other
+    @Transactional
+    public synchronized void acceptOwnMessage(final byte[] message) {
+
+        final EncodedPayload payload = payloadEncoder.decode(message);
+
+        final MessageHash transactionHash = Optional.of(payload)
+            .map(EncodedPayload::getCipherText)
+            .map(messageHashFactory::createFromCipherText)
+            .get();
+
+        final PublicKey sender = payload.getSenderKey();
+        if (!enclave.getPublicKeys().contains(sender)) {
+            throw new IllegalArgumentException(
+                "Message " + transactionHash.toString() + " does not have one the nodes own keys as a sender"
+            );
+        }
+
+        //this is a tx which we created
+        final Optional<EncryptedTransaction> tx = this.encryptedTransactionDAO.retrieveByHash(transactionHash);
+
+        if (tx.isPresent()) {
+
+            //we just need to add the recipient
+            final byte[] encodedPayload = tx.get().getEncodedPayload();
+            final EncodedPayload existing = payloadEncoder.decode(encodedPayload);
+
+            if (!existing.getRecipientKeys().contains(payload.getRecipientKeys().get(0))) {
+                //You could check that the provided information really does decrypt the payload
```

Comment:
I think this check should actually be implemented. It is not hard to see a "malicious" party which sends bogus garbage for every unknown transaction (not sure how easy it is to produce a byte array which when hashed results in a certain hash). In such a case that party would be added to the recipients list which would then allow it to invoke the resend and then gain access to the actual encoded payload - which would then facilitate a brute force attack.
The check should be that decrypting the box produces the expected master key, and that the decrypted transaction payload from the input transaction matches the decrypted  payload from the DB.

Reply:
I'm happy to add the check. I think a brute force attack would still be infeasible, but better to make it entirely impossible.

Reply:
Added a guard that checks the cipher text and decrypted payloads match.